### PR TITLE
instructions (8-11)

### PR DESCRIPTION
**Pull request by yossef-khaled**

### CODE REVIEW

File: control_unit.py [hunk 1]
```diff
@@ -67,22 +67,33 @@ def _STR(instruction):
...
 def _RTI(instruction):
-    """unused"""
-    pass  # TODO throw exception here.
+    raise Exception("instruction is not implemented !")
+    pass  
 
 
 def _NOT(instruction):
-    """bitwise not"""
+    DR = (_NOT(instruction) >> 9) & 0X7                            # Destenation Register.
```
Comment (mhashim6) [inline control_unit.py]: `(_NOT(instruction) >> 9)`
that's an endless recursive call.
You probably mean:
`DR = (instruction >> 9) & 0X7`

File: control_unit.py [hunk 2]
```diff
@@ -67,22 +67,33 @@ def _STR(instruction):
...
+    SR = (_NOT(instruction) >> 6) & 0X7                            # Source Register (register cntaining the data).
```
Comment (mhashim6) [inline control_unit.py]: same here, that's an endless recursive call.

File: control_unit.py [hunk 3]
```diff
@@ -67,22 +67,33 @@ def _STR(instruction):
...
+    for B in range (16):                                           # B stands (in this line) for Bit.
```
Comment (mhashim6) [inline control_unit.py]: it seems like you're `inverting` each bit separately, you don't need to do that, Python (like almost every other language) allows you to manipulate the variable at once with a single `~` expression:
`reg_write(Registers(DR), ~reg_read(Registers(SR)))`

File: control_unit.py [hunk 4]
```diff
@@ -67,22 +67,33 @@ def _STR(instruction):
...
+        reg_read(Registers(DR))[B] = ~reg_read(Registers(SR))[B]   # make every B in DR equals to the flip of the B in the same index in SR.
```
Comment (mhashim6) [inline control_unit.py]: btw, `Registers(DR)` is not a list, You can't access it with `[]`, and you don't need to, as I've mentioned in my previous comment.

File: control_unit.py [hunk 5]
```diff
@@ -67,22 +67,33 @@ def _STR(instruction):
...
+    SR = (_NOT(instruction) >> 6) & 0X7                            # Source Register (register cntaining the data).
+    for B in range (16):                                           # B stands (in this line) for Bit.
+        reg_read(Registers(DR))[B] = ~reg_read(Registers(SR))[B]   # make every B in DR equals to the flip of the B in the same index in SR.   
+    update_flags(DR)                                               # store the sign of the last excuted instruction data (which is in the DR).
     pass
 
 
 def _LDI(instruction):
-    """load indirect"""
+    DR = (_LDI(instruction) >> 9) & 0x7                            # Destenation Register.
```
Comment (mhashim6) [inline control_unit.py]: same here, that's an endless recursive call.

File: control_unit.py [hunk 6]
```diff
@@ -67,22 +67,33 @@ def _STR(instruction):
...
+    PCoffset = sign_extend(_LDI(instruction) & 0X1ff, 7)           # the value of what called an offset (embedded within the instruction code).
```
Comment (mhashim6) [inline control_unit.py]: same here, that's an endless recursive call.

Comment (mhashim6) [reply control_unit.py]: also, you should `sign_extend` with 9 bits, not 7.

File: control_unit.py [hunk 7]
```diff
@@ -67,22 +67,33 @@ def _STR(instruction):
...
+    address = reg_read(Registers.PC) + PCoffset                    # the address of the desired data.
+    reg_write(Registers(DR), mem_read(address))                    # write the data (stored in the address) in the register (DR).
```
Comment (mhashim6) [inline control_unit.py]: `LDI` stands for `load indirect`, this line should be:
`reg_write(Registers(DR), mem_read(mem_read(address)))`

File: control_unit.py [hunk 8]
```diff
@@ -67,22 +67,33 @@ def _STR(instruction):
...
+    update_flags(DR)                                               # store the sign of the last excuted instruction data (which is in the DR).
     pass
 
 
 def _STI(instruction):
-    """store indirect"""
+    SR = (_STI(instruction) >> 9) & 0x7                            # Source Register (the register containing the data).
+    PCoffset = sign_extend(_LDI(instruction) & 0xff1, 7)           # the value of what called an offset (embedded within the instruction code).
```
Comment (mhashim6) [inline control_unit.py]: endless recursive call.
Also, you should `sign_extend` with 9 bits, not 7.

Comment (mhashim6) [reply control_unit.py]: `0xff1` should be `0x1ff`

File: control_unit.py [hunk 9]
```diff
@@ -67,22 +67,33 @@ def _STR(instruction):
...
+    address = reg_read(Registers.PC) + PCoffset                    # the address that the data will be stored at.
+    mem_write(address, reg_read(Registers(SR)))                    # write the data (stored in the SR) into the memory (specifically at "address").
```
Comment (mhashim6) [inline control_unit.py]: `STI` stands for `store indirect`, this line should be:
`mem_write(mem_read(address), reg_read(Registers(SR)))`

File: control_unit.py [hunk 10]
```diff
@@ -67,22 +67,33 @@ def _STR(instruction):
...
     pass
```
Comment (mhashim6) [inline control_unit.py]: unnecessary `pass`

File: control_unit.py [hunk 11]
```diff
@@ -67,22 +67,33 @@ def _STR(instruction):
 
 
 def _RTI(instruction):
-    """unused"""
-    pass  # TODO throw exception here.
+    raise Exception("instruction is not implemented !")
+    pass
```
Comment (mhashim6) [inline control_unit.py]: unnecessary `pass`

File: control_unit.py [hunk 12]
```diff
@@ -67,22 +67,33 @@ def _STR(instruction):
...
 
 
 def _NOT(instruction):
-    """bitwise not"""
+    DR = (_NOT(instruction) >> 9) & 0X7                            # Destenation Register.
+    SR = (_NOT(instruction) >> 6) & 0X7                            # Source Register (register cntaining the data).
+    for B in range (16):                                           # B stands (in this line) for Bit.
+        reg_read(Registers(DR))[B] = ~reg_read(Registers(SR))[B]   # make every B in DR equals to the flip of the B in the same index in SR.   
+    update_flags(DR)                                               # store the sign of the last excuted instruction data (which is in the DR).
     pass
```
Comment (mhashim6) [inline control_unit.py]: unnecessary `pass`

File: control_unit.py [hunk 13]
```diff
@@ -67,22 +67,33 @@ def _STR(instruction):
...
 
 
 def _LDI(instruction):
-    """load indirect"""
+    DR = (_LDI(instruction) >> 9) & 0x7                            # Destenation Register.
+    PCoffset = sign_extend(_LDI(instruction) & 0X1ff, 7)           # the value of what called an offset (embedded within the instruction code).
+    address = reg_read(Registers.PC) + PCoffset                    # the address of the desired data.
+    reg_write(Registers(DR), mem_read(address))                    # write the data (stored in the address) in the register (DR).
+    update_flags(DR)                                               # store the sign of the last excuted instruction data (which is in the DR).
     pass
```
Comment (mhashim6) [inline control_unit.py]: unnecessary `pass`

File: control_unit.py (hunk 11 above)
Comment (yossef-khaled) [reply control_unit.py]: just forgot to delete the pass lines in every instruction

File: control_unit.py (hunk 1 above)
Comment (yossef-khaled) [reply control_unit.py]: wrote that line first as you did in the comment but i thought i did it wrong

File: control_unit.py (hunk 2 above)
Comment (yossef-khaled) [reply control_unit.py]: workin on it

File: control_unit.py (hunk 3 above)
Comment (yossef-khaled) [reply control_unit.py]: dig it

File: control_unit.py (hunk 4 above)
Comment (yossef-khaled) [reply control_unit.py]: isn't every register an array itself ?

File: control_unit.py (hunk 1 above)
Comment (mhashim6) [reply control_unit.py]: no problem, but I don't understand why did you think calling `_NOT(instruction)` is the right thing to do, can you elaborate?

File: control_unit.py (hunk 12 above)
Comment (yossef-khaled) [reply control_unit.py]: dig it

File: control_unit.py (hunk 5 above)
Comment (yossef-khaled) [reply control_unit.py]: dig it

File: control_unit.py (hunk 4 above)
Comment (mhashim6) [reply control_unit.py]: Nope, every register is an `enum` with a constant value that represents it's index in the internal registers array.

File: control_unit.py (hunk 6 above)
Comment (yossef-khaled) [reply control_unit.py]: thought that i pass the number of the rest bits (ones i want them to be added) to the sign_extend function , my bad

File: control_unit.py (hunk 7 above)
Comment (yossef-khaled) [reply control_unit.py]: dig it

File: control_unit.py (hunk 13 above)
Comment (yossef-khaled) [reply control_unit.py]: dig it

File: control_unit.py (hunk 8 above)
Comment (yossef-khaled) [reply control_unit.py]: converted them wrong to hex , my bad

File: control_unit.py (hunk 9 above)
Comment (yossef-khaled) [reply control_unit.py]: dig it

File: control_unit.py (hunk 10 above)
Comment (yossef-khaled) [reply control_unit.py]: dig it

File: control_unit.py (hunk 8 above)
Comment (mhashim6) [reply control_unit.py]: did you convert it manually?

File: control_unit.py [hunk 14]
```diff
@@ -67,23 +67,33 @@ def _STR(instruction):
...
-    """unused"""
-    pass  # TODO throw exception here.
+    raise Exception("instruction is not implemented !")
+    
 
 
 def _NOT(instruction):
-    """bitwise not"""
-    pass
+    DR = ((instruction) >> 9) & 0X7                                # Destenation Register.
```
Comment (mhashim6) [inline control_unit.py]:
```suggestion
    DR = (instruction >> 9) & 0X7                                # Destination Register.
```

File: control_unit.py [hunk 15]
```diff
@@ -67,23 +67,33 @@ def _STR(instruction):
...
-    pass  # TODO throw exception here.
+    raise Exception("instruction is not implemented !")
+    
 
 
 def _NOT(instruction):
-    """bitwise not"""
-    pass
+    DR = (instruction >> 9) & 0X7                                # Destination Register.
+    SR = ((instruction) >> 6) & 0X7                                # Source Register (register cntaining the data).
```
Comment (mhashim6) [inline control_unit.py]:
```suggestion
    SR = (instruction >> 6) & 0X7                                # Source Register (register cntaining the data).
```

File: control_unit.py [hunk 16]
```diff
@@ -67,23 +67,33 @@ def _STR(instruction):
...
+    SR = (instruction >> 6) & 0X7                                # Source Register (register cntaining the data).
+    reg_write(Registers(DR), ~reg_read(Registers(SR)))             # every bit in the DR will equal to the flipped one with the same index in SR. 
+    update_flags(DR)                                               # store the sign of the last excuted instruction data (which is in the DR).
+    
 
 
 def _LDI(instruction):
-    """load indirect"""
-    pass
+    DR = ((instruction) >> 9) & 0x7                                # Destenation Register.
```
Comment (mhashim6) [inline control_unit.py]:
```suggestion
    DR = (instruction >> 9) & 0x7                                # Destenation Register.
```

File: control_unit.py [hunk 17]
```diff
@@ -67,23 +67,33 @@ def _STR(instruction):
...
+    address = reg_read(Registers.PC) + PCoffset                    # the address of the address of the desired data.
+    reg_write(Registers(DR), mem_read(mem_read(address)))          # write the data (its address is explained in the previous line) in the register (DR).
+    update_flags(DR)                                               # store the sign of the last excuted instruction data (which is in the DR).
+    
 
 
 def _STI(instruction):
-    """store indirect"""
-    pass
+    SR = ((instruction) >> 9) & 0x7                                # Source Register (the register containing the data).
```
Comment (mhashim6) [inline control_unit.py]:
```suggestion
    SR = (instruction >> 9) & 0x7                                # Source Register (the register containing the data).
```

**Comment by mhashim6** [comment]
@yossef-khaled 
so far so good, now mark your instructions as checked in the `README.md` file.

**Comment by yossef-khaled** [comment]
ok

**Pull request merged**